docs/design/README.md

```diff
@@ -0,0 +1,595 @@
+- `variants`: A Cargo workspace for defining what Bottlerocket variants.
+
+A new directory will be added: `kits`.
+This will be a Cargo workspace with one Cargo package for each kit.
+
+```
+kits
+├── aws
+│   ├── build.rs
+│   ├── Cargo.toml
+│   └── lib.rs
+├── core
+│   ├── build.rs
+│   ├── Cargo.toml
+│   └── lib.rs
+├── ecs
+│   ├── build.rs
+│   ├── Cargo.toml
+│   └── lib.rs
+├── k8s
+│   ├── build.rs
+│   ├── Cargo.toml
+│   └── lib.rs
+├── metal
+│   ├── build.rs
+│   ├── Cargo.toml
+│   └── lib.rs
+└── vmware
+    ├── build.rs
+    ├── Cargo.toml
+    └── lib.rs
+```
+
+A variant such as `metal-k8s-1.22` would consume the necessary kits.
+For example, it would reference `core-kit`, `k8s-kit` and `metal-kit`.
+To add `aws-iam-authenticator`, `aws-kit` would be added.
+
+It may be convenient for certain build steps to have a single package representing a kit, such as `bottlerocket-aarch64-core-kit-0.1-1.aarch64.rpm`
+If so, these spec files could be added to the above directory tree.
+It is also possible to generate these spec files as they would be simple aggregations of all the packages in the kit.
+
+### Kit Structure
+
+A kit container will be built using `FROM scratch` because it does not need a runtime environment.
+It's directory structure will look like this:
+
+```sh
+# tree /local --dirsfirst
+/local
+├── etc
+│   └── yum.repos.d
+│       └── bottlerocket-core-kit.repo
+└── kits
+    └── bottlerocket-core-kit
+        ├── archives
+        │   └── bottlerocket-x86_64-kernel-5.15-archive-5.15.102-1.x86_64.rpm
+        ├── repodata
+        │   ├── 30e098715e13cbe4733c6b4518dcc2cdc68d1ac3277eefad2ee5ac171c4b9023-primary.sqlite.bz2
+        │   ├── 5a22c22832fa81e55f84de715cf910cebc64d472be65a7124fd8b02ff7c777fb-filelists.xml.gz
+        │   ├── 719d1eef5ce4d194d4c5c51171dce8987cb21e88338f1d5dff7554dacabf1004-other.sqlite.bz2
+        │   ├── 7cbab1a2f05c1eda86d3d125497af680d9100665e3eb30c7560926eabbc2c655-primary.xml.gz
+        │   ├── d61fe62ca166eab1cbd502cf21cd977798ef348de49fb35e38a93ebf696eb0a5-filelists.sqlite.bz2
+        │   ├── fcd4dadc59c6d6512410b103a38762bf266389e01a60add12319191c4cc16ec4-other.xml.gz
+        │   └── repomd.xml
+        ├── bottlerocket-x86_64-acpid-2.0.34-1.x86_64.rpm
+        ├── bottlerocket-x86_64-apiclient-0.0-0.x86_64.rpm
+        ├── ...
+        └── bottlerocket-x86_64-wicked-0.6.68-1.x86_64.rpm
+```
+
+The `/etc/yum.repos.d` file can be used later to configure `dnf`.
+This will allow `dnf` to treat each kit as a yum repository.
+When aggregating kits, the yum repos will be prioritized according to the maintainer's requirements.
+
+### Kit Dependencies
+
+One kit may depend on a package from another kit.
+All kits used together, need to have been built by the same SDK.
+Another way to say this is that all kits must depend on the same SDK.
+These dependencies need to be specified at the kit level so that tooling understands which kits are required for a build.
```

Comment:
The SDK is the starting point for this problem, given that Rust binaries are aggressively linked to the `libstd-*.so` shared library.

However, it can arise for kit:kit dependencies as well, for essentially the same reason: any shared library that one kit contributes can become a hard dependency for its downstream kits.